examples/quantize/quantize.cpp

```diff
@@ -26,6 +26,8 @@ static const std::vector<struct quant_option> QUANT_OPTIONS = {
     { "IQ2_M",  LLAMA_FTYPE_MOSTLY_IQ2_M,  " 2.7  bpw quantization",            },
     { "IQ1_S",  LLAMA_FTYPE_MOSTLY_IQ1_S,  " 1.56 bpw quantization",            },
     { "IQ1_M",  LLAMA_FTYPE_MOSTLY_IQ1_M,  " 1.75 bpw quantization",            },
+    { "Q1_3",   LLAMA_FTYPE_MOSTLY_Q1_3,   " 1.63 bpw for BitNet b1.58",        },
+    { "Q2_2",   LLAMA_FTYPE_MOSTLY_Q2_2,   " 2.00 bpw for BitNet b1.58",        },
```

Comment:
Regarding the names of the new quant types, since these are quite specific to BitNet models, I was thinking to name them with something starting with `QB`, a bit like suggested in <https://github.com/ggerganov/llama.cpp/issues/5761#issuecomment-2071270724>.

I'll first be describing what I want from the naming scheme, then I'll attempt to make it work.

The naming scheme should have room for:
- Ternary types in `{-1, 0, 1}`
  - `1.625 bpw` quant with a block size of 64, with 13 bytes per block
    - To make the smallest possible lossless BitNet b1.58 model files
    - Uses `Q8_0` as its `vec_dot_type` (for the activations)
    - (It's technically possible to store a `float16` scale in the leftover bits in the last byte of 16 consecutive blocks (this means 1024 elements minimum per row), although it can't really be extracted with SIMD)
  - `2.000 bpw` quant with a block size of 32, with 8 bytes per block
    - For maximal performance
    - Uses `Q8_0` as its `vec_dot_type` (for the activations)
  - `2.000 bpw` quant with a block size of 64, with 16 bytes per block, and a float16 scale
    - Values would be packed similarly to the `1.625 bpw` type, but with an extra byte and a row-wise `float16` scale duplicated in each block.
  - `2.000 bpw` quant with a block size of 4, with 1 byte per block
    - For weirdly-shaped models like the 1.3B BitNet b1.58 model
    - Needs a compatible `vec_dot_type`
      - float types are slower than integer types for this
- Binary types in `{-1, 1}`
  - `1 bpw` type
- Binary types in `{0, 1}`
  - Are there models which use this?
- 8-bit activation with a row-wise scale
  - `8.5 bpw` like `Q8_0`, but all the scales of a row are the same
    - Would allow reducing the number of `float32` operations in the vec_dot of the above types.
  - `10 bpw`, 5 bytes per block of 4 elements, with a weird layout which only uses blocks to get a big enough buffer, with a single float32 scale and some padding before all row elements, aligned and contiguous.
    - For use with the weird `2.000 bpw` type, and also maybe the other ones for best performance.

So the naming scheme could be:

- `QB<x>_<y>`
  - where `<x>` is the floor of the expected bpw of the type
  - where `<y>` is
    - `0` binary type, `{0, 1}`
      - except for `QB8_0` which is like `Q8_0` but with a guaranteed duplicated row-wise scale
    - `1` binary type, `{-1, 1}`
    - `2` ternary type using some kind of binary-coded ternary
    - `3` ternary type with fixed-point packed values
    - `4` weird type with a block size of 4

Which for the previously-mentioned possible BitNet types would mean:

| proposed name | Range | bits per weight | block size | bytes | row-wise scale | current name |
|-|-|-|-|-|-|-|
|`QB1_3`| `{-1, 0, 1}` | 1.625 | 64 | 13 | 1.0f | `Q1_3` |
|`QB2_2`| `{-2, -1, 0, 1}` | 2.000 | 32 | 8 | 1.0f | `Q2_2` |
|`QB2_3`| `{-1, 0, 1}` | 2.000 | 64 | 16 | `f16` | |
|`QB2_4`| `{-2, -1, 0, 1}` | 2.000 | 4 | 1 | 1.0f | |
|`QB1_1`| `{-1, 1}` | 1.000 | ? | ?/8 | 1.0f | |
|`QB1_0`| `{0, 1}` | 1.000 | ? | ?/8 | 1.0f | |
|`QB8_0`| `[-127, 127]` | 8.5 | 32 | 34 | `f16` | |
|`QB8_4`| `[-127, 127]` | 10 | 4 | 5 | `f32`, weird layout | |

I'm not saying these should all exist, though, only that the naming scheme should not be too limiting for possible future extensions.

So I think I'll rename `Q1_3` to `QB1_3`, and `Q2_2` to `QB2_2`. Anyone has comments on this? Or a better naming scheme for the new BitNet quant types?

Reply:
If it were me, considering this only works with bitnet models and nothing else, I'd want the designations to be exceptionally clear that they are different and shouldn't be used on just anything.  "QB" is good, but I'd take it a step further and remove the Q entirely.  As bitnet is being colloquially referred to as a "1-bit" model, B1 makes more sense.  Considering the plausible range for weights, I'd cut it off at tenths and ditch the decimal.  This leaves plenty of room for variations, while making the native BPW very clear.  I feel this is superior to the arbitrary "_2" and "_3" subtypes.

So what I would propose is:

1.625bpw = `B1_16`
2.000bpw = `B1_20`